pyannote/audio/augmentation/noise.py

```diff
@@ -127,5 +127,6 @@ def __call__(self, original, sample_rate):
         # select SNR at random
         snr = (self.snr_max - self.snr_min) * np.random.random_sample() + self.snr_min
         alpha = np.exp(-np.log(10) * snr / 20)
-
+        if len(noise) > len(original):
+            noise = noise[:len(original)]
```

Comment:
Is this really necessary?

My guess is that you had to add this check because data augmentation failed when combining NIST SRE data (8kHz) with MUSAN data (16kHz). Am I correct?

This is likely due to the way that start/end indices in `RawAudio.crop`  are modified in the case of on-the-fly resampling (at lines 278 and 279).

We should make sure that `RawAudio.crop` returns the correct number of samples **in the target sample rate** (as defined at line 264).

Reply:
I had to add this check for the situation when either original or noise data are resampled (you are correct) and also when noise comes from several smaller segments. If the noise comes from one segment, there is not mismatch, it is correctly compute, but the `len(noise)` becomes `len(original)+1`, when noise is concatenated from several segments. There is a rounding error somewhere.

Reply:
Computing duration of the noise based on the returned signal than what is in the file header addresses this issue. Check commit d16cae6

Reply:
This currently breaks the semantics of (undocumented, I know...) `fixed` parameter in `RawAudio.crop`. 

Two calls to `RawAudio.crop` with the same `fixed` value should return the same number of samples. This is no longer true if two files have two different sampling rates.

Let me have a look at this (and also the +1 rounding error).

Reply:
Is it possible to have different ``self.sample_rate`` in the ``RawAudio.crop``? As I understand it, the ``self.sample_rate`` comes from the ``config.yml`` file, so it is hardcoded and hence calling ``RawAudio.crop`` with the same `fixed` parameters (it's only used here: https://github.com/pyannote/pyannote-audio/blob/cf2dde9887bc9c5e2e179ed066f5ae91e55feaf6/pyannote/audio/features/utils.py#L264) should return the same number of samples.

Reply:
As for the rounding error, there are two sources for this error. 
1. If `self.sample_rate` (say 8000) is smaller than `sample_rate` (say 16000), then the problem occurs in `noise.py` when the `duration=0.0076875` of the noise utterance in seconds is qual to `123 samples` given the sample rate of `16000`. When downsampled, it should become `61.5 samples` but it is rounded up to `62`  by `librosa`, leading to the round error.
2. If `self.sample_rate` (16000) is larger than `sample_rate` (8000), then upsampling is done in `RawAudio.crop`.  After upsampling (at least from 8000 to 16000), the number of resulted samples is always an even number, since we double the number of samples for the same duration. It will lead to different number samples for the same duration in seconds, given different sample rates.

Reply:
`self.sample_rate` won't change, but `sample_rate` may change from file to file.

Reply:
I just pushed https://github.com/pyannote/pyannote-audio/commit/6275b7c5154f877dc999c44eb5132a8bc2ffbb18 that fixes it (I think).

I couldn't find a way to push directly to your pull request so pushed to a pkorshunov-develop branch instead. Are you able to incorporate this commit into your own pull request?

Reply:
This was my point that `self.sliding_window_.crop(segment, mode=mode, fixed=fixed, return_ranges=True)` is called without `sample_rate` as a parameter, so it is always computed using `self.sample_rate`. It means no matter what is the actual `sample_rate` of the file, the `self.sliding_window_.crop` will always return the same number of samples with the `fixed=True` option. Am I wrong?

Reply:
Yes. `self.sliding_window_.crop` will always return the same number of samples when `fixed` is given a value... and so should `RawAudio.crop`! 

However, because `start` and `end` are modified in the case of sample rate mismatch, it is no longer the case. Hence `RawAudio.crop` is broken.

Reply:
> I just pushed [6275b7c](https://github.com/pyannote/pyannote-audio/commit/6275b7c5154f877dc999c44eb5132a8bc2ffbb18) that fixes it (I think).
> 
> I couldn't find a way to push directly to your pull request so pushed to a pkorshunov-develop branch instead. Are you able to incorporate this commit into your own pull request?

I managed to update your pull request directly...